finality-aleph/src/session_map.rs

```diff
@@ -281,76 +264,95 @@ where
+        } else {
+            panic!("Authorities for next session {:?} must be available at first block #{:?} of current session", session_id.0, first_block);
         }
 
-        if session_id.0 >= PRUNING_THRESHOLD && session_id.0 % PRUNING_THRESHOLD == 0 {
-            debug!(target: "aleph-session-updater", "Pruning session map below session #{:?}", session_id.0 - PRUNING_THRESHOLD);
+        if session_id.0 > PRUNING_THRESHOLD && session_id.0 % PRUNING_THRESHOLD == 0 {
+            debug!(target: "aleph-session-updater",
+                "Pruning session map below session #{:?}",
+                session_id.0 - PRUNING_THRESHOLD
+            );
             self.session_map
                 .prune_below(SessionId(session_id.0 - PRUNING_THRESHOLD))
                 .await;
         }
     }
 
-    async fn update_session(&mut self, session_id: SessionId, period: SessionPeriod) {
-        let first_block = first_block_of_session(session_id, period);
-        self.handle_first_block_of_session(first_block, session_id)
-            .await;
+    fn authorities_for_session(&mut self, session_id: SessionId) -> Option<SessionAuthorityData> {
+        let first_block = first_block_of_session(session_id, self.period);
+        self.authority_provider.authority_data(first_block)
     }
 
-    fn catch_up_boundaries(&self, period: SessionPeriod) -> (SessionId, SessionId) {
+    /// Puts current and next session authorities in the session map.
+    /// If previous authorities are still available in `AuthorityProvider`, also puts them in the session map.
+    async fn catch_up(&mut self) -> SessionId {
         let last_finalized = self.finality_notificator.last_finalized();
 
-        let current_session = session_id_from_block_num(last_finalized, period);
-        let starting_session = SessionId(current_session.0.saturating_sub(PRUNING_THRESHOLD));
+        let current_session = session_id_from_block_num(last_finalized, self.period);
+        let starting_session = SessionId(current_session.0.saturating_sub(PRUNING_THRESHOLD - 1));
 
-        (starting_session, current_session)
-    }
+        debug!(target: "aleph-session-updater",
+            "Last finalized is {:?}; Catching up with authorities starting from session {:?} up to next session {:?}",
+            last_finalized, starting_session.0, current_session.0 + 1
+        );
 
-    pub async fn run(mut self, period: SessionPeriod) {
-        let mut notifications = self.finality_notificator.notification_stream();
+        // lets catch up with previous sessions
+        for session in starting_session.0..current_session.0 {
+            let id = SessionId(session);
+            if let Some(authority_data) = self.authorities_for_session(id) {
+                self.session_map.update(id, authority_data).await;
+            } else {
+                debug!(target: "aleph-session-updater", "No authorities for session {:?} during catch-up. Most likely already pruned.", id.0)
```

Comment:
Im pretty sure at this point we should already panic. If we don't know the authorities for some older sessions how can we make sure the current session authorities are correct

Reply:
right, good point